frame/nis/src/lib.rs

```diff
@@ -0,0 +1,914 @@
+						T::QueueCount::get(),
+						u32::max_value(),
+						&mut weight_counter,
+					)
+				}
+			}
+			weight_counter.used
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Place a bid.
+		///
+		/// Origin must be Signed, and account must have at least `amount` in free balance.
+		///
+		/// - `amount`: The amount of the bid; these funds will be reserved, and if/when
+		///   consolidated, removed. Must be at least `MinBid`.
+		/// - `duration`: The number of periods before which the newly consolidated bid may be
+		///   thawed. Must be greater than 1 and no more than `QueueCount`.
+		///
+		/// Complexities:
+		/// - `Queues[duration].len()` (just take max).
+		#[pallet::weight(T::WeightInfo::place_bid_max())]
+		pub fn place_bid(
+			origin: OriginFor<T>,
+			#[pallet::compact] amount: BalanceOf<T>,
+			duration: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(amount >= T::MinBid::get(), Error::<T>::AmountTooSmall);
+			let queue_count = T::QueueCount::get() as usize;
+			let queue_index = duration.checked_sub(1).ok_or(Error::<T>::DurationTooSmall)? as usize;
+			ensure!(queue_index < queue_count, Error::<T>::DurationTooBig);
+
+			let net = Queues::<T>::try_mutate(
+				duration,
+				|q| -> Result<(u32, BalanceOf<T>), DispatchError> {
+					let queue_full = q.len() == T::MaxQueueLen::get() as usize;
+					// ensure!(have_reserved(q[0].amount));
+					ensure!(!queue_full || q[0].amount < amount, Error::<T>::BidTooLow);
+					T::Currency::reserve(&who, amount)?;
+
+					// queue is <Ordered: Lowest ... Highest><Fifo: Last ... First>
+					let mut bid = Bid { amount, who: who.clone() };
+					let net = if queue_full {
+						sp_std::mem::swap(&mut q[0], &mut bid);
+						let _ = T::Currency::unreserve(&bid.who, bid.amount);
+						(0, amount - bid.amount)
+					} else {
+						q.try_insert(0, bid).expect("verified queue was not full above. qed.");
+						(1, amount)
+					};
+
+					let sorted_item_count = q.len().saturating_sub(T::FifoQueueLen::get() as usize);
+					if sorted_item_count > 1 {
+						q[0..sorted_item_count].sort_by_key(|x| x.amount);
+					}
+
+					Ok(net)
+				},
+			)?;
+			QueueTotals::<T>::mutate(|qs| {
+				qs.bounded_resize(queue_count, (0, Zero::zero()));
+				qs[queue_index].0 += net.0;
+				qs[queue_index].1 = qs[queue_index].1.saturating_add(net.1);
+			});
+			Self::deposit_event(Event::BidPlaced { who, amount, duration });
+
+			Ok(())
+		}
+
+		/// Retract a previously placed bid.
+		///
+		/// Origin must be Signed, and the account should have previously issued a still-active bid
+		/// of `amount` for `duration`.
+		///
+		/// - `amount`: The amount of the previous bid.
+		/// - `duration`: The duration of the previous bid.
+		#[pallet::weight(T::WeightInfo::place_bid(T::MaxQueueLen::get()))]
+		pub fn retract_bid(
+			origin: OriginFor<T>,
+			#[pallet::compact] amount: BalanceOf<T>,
+			duration: u32,
+		) -> DispatchResultWithPostInfo {
```

Comment:
```suggestion
		) -> DispatchResult {
```
I dont see it being utilized.